wgpu-hal/src/dx12/sampler.rs

```diff
@@ -0,0 +1,251 @@
+#[derive(Debug, Copy, Clone)]
+struct HashableSamplerDesc(D3D12_SAMPLER_DESC);
+
+impl PartialEq for HashableSamplerDesc {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.Filter == other.0.Filter
+            && self.0.AddressU == other.0.AddressU
+            && self.0.AddressV == other.0.AddressV
+            && self.0.AddressW == other.0.AddressW
+            && OrderedFloat(self.0.MipLODBias) == OrderedFloat(other.0.MipLODBias)
+            && self.0.MaxAnisotropy == other.0.MaxAnisotropy
+            && self.0.ComparisonFunc == other.0.ComparisonFunc
+            && self.0.BorderColor.map(OrderedFloat) == other.0.BorderColor.map(OrderedFloat)
+            && OrderedFloat(self.0.MinLOD) == OrderedFloat(other.0.MinLOD)
+            && OrderedFloat(self.0.MaxLOD) == OrderedFloat(other.0.MaxLOD)
+    }
+}
+
+impl Eq for HashableSamplerDesc {}
+
+impl std::hash::Hash for HashableSamplerDesc {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.Filter.0.hash(state);
+        self.0.AddressU.0.hash(state);
+        self.0.AddressV.0.hash(state);
+        self.0.AddressW.0.hash(state);
+        OrderedFloat(self.0.MipLODBias).hash(state);
+        self.0.MaxAnisotropy.hash(state);
+        self.0.ComparisonFunc.0.hash(state);
+        self.0.BorderColor.map(OrderedFloat).hash(state);
+        OrderedFloat(self.0.MinLOD).hash(state);
+        OrderedFloat(self.0.MaxLOD).hash(state);
+    }
+}
+
+/// Entry in the sampler cache.
+struct CacheEntry {
+    index: SamplerIndex,
+    ref_count: u32,
+}
+
+/// Container for the mutable management state of the sampler heap.
+///
+/// We have this separated, using interior mutability, to allow for the outside world
+/// to access the heap directly without needing to take the lock.
+pub(crate) struct SamplerHeapState {
+    /// Mapping from the sampler description to the index within the heap and the refcount.
+    mapping: HashMap<HashableSamplerDesc, CacheEntry>,
+    /// List of free sampler indices.
+    freelist: Vec<SamplerIndex>,
+}
+
+/// Global sampler heap for the device.
+///
+/// As D3D12 only allows 2048 samplers to be in a single heap, we need to cache
+/// samplers aggressively and refer to them in shaders by index.
+pub(crate) struct SamplerHeap {
+    /// Mutable management state of the sampler heap.
+    state: Mutex<SamplerHeapState>,
+
+    /// The heap itself.
+    heap: ID3D12DescriptorHeap,
+    /// The CPU-side handle to the first descriptor in the heap.
+    ///
+    /// Both the CPU and GPU handles point to the same descriptor, just in
+    /// different contexts.
+    heap_cpu_start_handle: D3D12_CPU_DESCRIPTOR_HANDLE,
+    /// The GPU-side handle to the first descriptor in the heap.
+    ///
+    /// Both the CPU and GPU handles point to the same descriptor, just in
+    /// different contexts.
+    heap_gpu_start_handle: D3D12_GPU_DESCRIPTOR_HANDLE,
+
+    /// This is the device-specific size of sampler descriptors.
+    heap_stride: u32,
```

Comment:
I was confused about the name of this, how about `descriptor_stride`?

Reply:
Sure